### PR TITLE
Enable tag completion in front matter

### PR DESCRIPTION
**Pull request by jimgraham**
Addresses #1184

Currently tag completion only works in the front matter if you type a `#` character. Adding the suggested tag will then mark the tag as a comment

```markdown
---
tags: #foo #bar
---
```

Update the tag completion provider to recognize if we are in the front-matter. Because the tag completion intellisense must be summoned with either a `#` character or the keybinding (typically `ctrl+space`), allow for 2 outcomes

1. if the tag is prefixed in the front matter with a `#`, remove it when substituting the tag.
2. If `ctrl-space` is used, recognize we are on the `tags: ` line and allow for non-`#` prefaced words.

The tag provide only works on the `tags: `section of the front-matter. 

## Update

I believe I've addressed all the feedback

Here's a video in action: 

https://user-images.githubusercontent.com/430293/232338426-0b8acf57-df36-4e50-a2ef-d6e6307c7baa.mp4

### CODE REVIEW

File: packages/foam-vscode/src/features/tag-completion.ts [hunk 1]
```diff
@@ -1,10 +1,16 @@
...
+import matter from 'gray-matter';
 import { Foam } from '../core/model/foam';
 import { FoamTags } from '../core/model/tags';
 import { HASHTAG_REGEX } from '../core/utils/hashtags';
 import { FoamFeature } from '../types';
 import { mdDocSelector } from '../utils';
 
+const MAX_FRONT_MATTER_LINES = 50;
+const FRONT_MATTER_TAG_REGEX =
+  /^\s*tags:\s+.*?\s{0,1}?#?([0-9]*[\p{L}\p{Emoji_Presentation}/_-][\p{L}\p{Emoji_Presentation}\p{N}/_-]*)$/gmu;
```
Comment (riccardoferretti) [inline packages/foam-vscode/src/features/tag-completion.ts:12]: will this work for all the tags or just the first one? also there is the question of frontmatter "flexibility", e.g. the following is valid frontmatter that creates 5 tags:

```
---
tags: aaa aai, aab
  aae
  aao
---
```

This is also valid YAML

```
---
tags: 
 - aaa aai, aab
  aae
  aao
 - aau
---
```

I would be ok with only supporting "same line" tags (seems like the majority of the cases imo), but we need to support the space and comma separators

**Review by riccardoferretti — COMMENTED** [review]
I have left a few comments, let me know what you think

File: packages/foam-vscode/src/features/tag-completion.ts [hunk 2]
```diff
@@ -55,6 +84,27 @@ export class TagCompletionProvider
...
+    document: vscode.TextDocument,
+    position: vscode.Position
+  ): Boolean {
+    // if the line is > 50, assume we're not in the front matter
+    if (position.line > MAX_FRONT_MATTER_LINES) {
+      return false;
+    }
+
+    const fm = matter(document.getText());
+    if ('tags' in fm.data) {
```
[the remark this answers is not in the document]
Comment (jimgraham) [reply packages/foam-vscode/src/features/tag-completion.ts]: I moved this to a `isInFrontMatter` utility function (with unit tests) in https://github.com/foambubble/foam/pull/1191/files#diff-2f15bd61b55d617a33f7675aa163c5ed8e168efd6b92630c0f5f16f37f7c470bR225

File: packages/foam-vscode/src/features/tag-completion.ts [hunk 3]
```diff
@@ -55,6 +84,27 @@ export class TagCompletionProvider
...
+  static isInFrontMatter(
+    document: vscode.TextDocument,
+    position: vscode.Position
+  ): Boolean {
+    // if the line is > 50, assume we're not in the front matter
+    if (position.line > MAX_FRONT_MATTER_LINES) {
+      return false;
+    }
+
+    const fm = matter(document.getText());
```
[the remark this answers is not in the document]
Comment (jimgraham) [reply packages/foam-vscode/src/features/tag-completion.ts]: I moved this to a `isInFrontMatter` utility function (with unit tests) in https://github.com/foambubble/foam/pull/1191/files#diff-2f15bd61b55d617a33f7675aa163c5ed8e168efd6b92630c0f5f16f37f7c470bR225

File: packages/foam-vscode/src/features/tag-completion.ts (hunk 1 above)
Comment (jimgraham) [reply packages/foam-vscode/src/features/tag-completion.ts]: checked in https://github.com/foambubble/foam/pull/1191/files#diff-2f15bd61b55d617a33f7675aa163c5ed8e168efd6b92630c0f5f16f37f7c470bR247

**Comment by jimgraham** [comment]
I'm getting some weird results where the completion is being triggered by non `#` characters like `-:`. I need to keep debugging. This PR is not yet ready for prime time. I will keep as draft until I figure it out.

thanks for your input.

**Comment by jimgraham** [comment]
@riccardoferretti This is ready for review. I understand that you may want to look at it in more depth given the last PR I shipped.

File: docs/user/features/tags.md [hunk 4]
```diff
@@ -49,7 +53,5 @@ The end result will be a CSS file that looks similiar to the content below. Now
 Given the power of backlinks, some people prefer to use them as tags.
 For example you can tag your notes about books with [[book]].
 
-[//begin]: # "Autogenerated link references for markdown compatibility"
```
Comment (jimgraham) [inline docs/user/features/tags.md:52]: not sure why this was "auto-formatted" away.

File: packages/foam-vscode/src/features/tag-completion.spec.ts [hunk 5]
```diff
@@ -213,4 +213,154 @@ more text
...
+      const tags = await provider.provideCompletionItems(
+        doc,
+        new vscode.Position(2, 11)
+      );
+
+      expect(foamTags.tags.get('primary')).toBeTruthy();
+      expect(tags).toBeNull();
+    });
+
+    it('should not provide suggestions when outside the front-matter without `#` key #1184', async () => {
```
Comment (jimgraham) [inline packages/foam-vscode/src/features/tag-completion.spec.ts:336]: should remove the issue number given it's in the `describe` block

File: packages/foam-vscode/src/features/tag-completion.ts [hunk 6]
```diff
@@ -38,30 +39,68 @@ export class TagCompletionProvider
       .lineAt(position)
       .text.substr(0, position.character);
 
-    const requiresAutocomplete = cursorPrefix.match(TAG_REGEX);
-    if (!requiresAutocomplete) {
+    const beginningOfFileText = document.getText(
+      new vscode.Range(
+        new vscode.Position(0, 0),
+        new vscode.Position(
+          position.line < 50 ? position.line : 50,
```
Comment (jimgraham) [inline packages/foam-vscode/src/features/tag-completion.ts:46]: can make `50` a const if necessary.

**Comment by jimgraham** [comment]
Yes, will do

**Comment by jimgraham** [comment]
> Do you think it would be possible to separate the two cases in different functions (in the end still called from the main fn)? 

@riccardoferretti I've refactored this code. Tests pass and I've locally tested it. Let me know what you think. Thanks

**Review by riccardoferretti — APPROVED** [review]
LGTM!

**Comment by jimgraham** [comment]
@riccardoferretti Anything you need from me to get this merged?

**Comment by riccardoferretti** [comment]
Oops, I thought I had already merged it!

**Pull request merged**